File: app/src/main/java/ir/apptune/calendar/ResourceUtils.java
package ir.apptune.calendar;

import android.content.Context;
import android.content.res.XmlResourceParser;
import android.util.Log;

import org.xmlpull.v1.XmlPullParser;

import java.util.HashMap;


/**
 * This class returns Persian Calendar events , with parsing the xml's inside the app.
 * please look at xml folder
 */

public final class ResourceUtils {
    private final static String XCalendarH = "ObservedHijriCalendar";
    private final static String XCalendarG = "GregorianCalendar";
    private final static String XCalendarP = "PersianCalendar";
    static HashMap<Integer, String> eventG, eventH, eventP;
    static HashMap<Integer, Boolean> vacationG;
    static HashMap<Integer, Boolean> vacationH;
    public static HashMap<Integer, Boolean> vacationP;

    public ResourceUtils(Context c) {
        eventP = new HashMap<Integer, String>();
        eventH = new HashMap<Integer, String>();
        eventG = new HashMap<Integer, String>();
        vacationP = new HashMap<Integer, Boolean>();
        vacationH = new HashMap<Integer, Boolean>();
        vacationG = new HashMap<Integer, Boolean>();
        getHashMapResource(c, R.xml.events_gregorian);
        getHashMapResource(c, R.xml.events_persian);
        getHashMapResource(c, R.xml.events_misc);
    }

    private static void getHashMapResource(Context c, int hashMapResId) {

        XmlResourceParser parser = c.getResources().getXml(hashMapResId);

        String key = null, value = null;
        String title, day, month, isVacation, xCalendar;

        try {
            int eventType = parser.getEventType();

            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_DOCUMENT) {
                    Log.d("utils", "Start document");
                } else if (eventType == XmlPullParser.START_TAG) {
                    if (parser.getName().equals("XCalendarEvents")) {


                    } else if (parser.getName().equals("Event")) {
                        title = parser.getAttributeValue(null, "Title");
                        day = parser.getAttributeValue(null, "Day");

                        month = parser.getAttributeValue(null, "Month");

                        isVacation = parser.getAttributeValue(null, "IsVacation");


                        xCalendar = parser.getAttributeValue(null, "XCalendar");
                        if (null == title) {
                            parser.close();
                            return;
                        }
                        if (day != null && month != null && xCalendar != null) {
                            int dayMonth = Integer.parseInt(month) * 100 + Integer.parseInt(day);
                            boolean vacation = false;
                            if (isVacation != null) if (isVacation.equals("1")) vacation = true;


                            if (xCalendar.equals(XCalendarP)) {
                                if (eventP.containsKey(dayMonth))
                                    eventP.put(dayMonth, eventP.get(dayMonth) + " " + title);
                                else
                                    eventP.put(dayMonth, title);
                                if (vacation) vacationP.put(dayMonth, true);
                            } else if (xCalendar.equals(XCalendarH)) {
                                if (eventH.containsKey(dayMonth))
                                    eventH.put(dayMonth, eventH.get(dayMonth) + " " + title);
                                else
                                    eventH.put(dayMonth, title);
                                if (vacation) vacationH.put(dayMonth, true);
                            } else if (xCalendar.equals(XCalendarG)) {
                                if (eventG.containsKey(dayMonth))
                                    eventG.put(dayMonth, eventG.get(dayMonth) + " " + title);
                                else

                                    eventG.put(dayMonth, title);
                                if (vacation) vacationG.put(dayMonth, true);

                            }
                        }
                    }
                } else if (eventType == XmlPullParser.END_TAG) {
                    if (parser.getName().equals("Event")) {


                        key = null;
                        value = null;
                    }
                } else if (eventType == XmlPullParser.TEXT) {
                    if (null != key) {
                        value = parser.getText();
                    }
                }
                eventType = parser.next();
            }
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }


    }

}

File: app/src/main/java/ir/apptune/calendar/ResourceUtils.kt
package ir.apptune.calendar

import android.content.Context
import android.util.Log
import org.koin.java.KoinJavaComponent
import org.xmlpull.v1.XmlPullParser

/**
 * This class returns Persian Calendar events , with parsing the xml's inside the app.
 * please look at xml folder
 */

private const val XCalendarH = "ObservedHijriCalendar"
private const val XCalendarG = "GregorianCalendar"
private const val XCalendarP = "PersianCalendar"
private const val EVENT = "Event"
private const val TITLE = "Title"
private const val DAY = "Day"
private const val MONTH = "Month"
private const val VACATION = "IsVacation"
private const val X_CALENDAR = "XCalendar"
private const val IS_VACATION = "1"

class ResourceUtils(c: Context) {

    private fun getHashMapResource(c: Context, hashMapResId: Int) {
        val parser = c.resources.getXml(hashMapResId)
        var key: String? = null
        var value: String? = null
        var title: String?
        var day: String?
        var month: String?
        var isVacation: String?
        var xCalendar: String?
        try {
            var eventType = parser.eventType
            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_DOCUMENT) {
                    Log.d("utils", "Start document")
                } else if (eventType == XmlPullParser.START_TAG) {
                    if (parser.name == EVENT) {
                        title = parser.getAttributeValue(null, TITLE)
                        day = parser.getAttributeValue(null, DAY)
                        month = parser.getAttributeValue(null, MONTH)
                        isVacation = parser.getAttributeValue(null, VACATION)
                        xCalendar = parser.getAttributeValue(null, X_CALENDAR)
                        if (null == title) {
                            parser.close()
                            return
                        }
                        if (day != null && month != null && xCalendar != null) {
                            val dayMonth = month.toInt() * 100 + day.toInt()
                            var vacation = false
                            if (isVacation != null) if (isVacation == IS_VACATION) vacation = true
                            if (xCalendar == XCalendarP) {
                                if (eventP.containsKey(dayMonth)) eventP[dayMonth] = eventP[dayMonth].toString() + " " + title else eventP[dayMonth] = title
                                if (vacation) vacationP[dayMonth] = true
                            } else if (xCalendar == XCalendarH) {
                                if (eventH.containsKey(dayMonth)) eventH[dayMonth] = eventH[dayMonth].toString() + " " + title else eventH[dayMonth] = title
                                if (vacation) vacationH[dayMonth] = true
                            } else if (xCalendar == XCalendarG) {
                                if (eventG.containsKey(dayMonth)) eventG[dayMonth] = eventG[dayMonth].toString() + " " + title else eventG[dayMonth] = title
                                if (vacation) vacationG[dayMonth] = true
                            }
                        }
                    }
                } else if (eventType == XmlPullParser.END_TAG) {
                    if (parser.name == EVENT) {
                        key = null
                        value = null
                    }
                } else if (eventType == XmlPullParser.TEXT) {
                    if (null != key) {
                        value = parser.text
                    }
                }
                eventType = parser.next()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return
        }
    }

    companion object {
        var eventG = hashMapOf<Int, String>()
        var eventH = hashMapOf<Int, String>()
        var eventP = hashMapOf<Int, String>()
        var vacationG = hashMapOf<Int, Boolean>()
        var vacationH = hashMapOf<Int, Boolean>()
        var vacationP = hashMapOf<Int, Boolean>()
    }

    init {
        getHashMapResource(c, R.xml.events_gregorian)
        getHashMapResource(c, R.xml.events_persian)
        getHashMapResource(c, R.xml.events_misc)
        getHashMapResource(c, R.xml.events_arabic)
    }
}